pre-check/v1.0.3/check.sh

```diff
@@ -0,0 +1,306 @@
+
+    if [ "$pending" = "false" ]; then
+      echo "All Helm bundles are ready."
+      return
+    fi
+
+    echo "There are non-ready Helm bundles:"
+    echo "$bundles" | yq '.items[] | select(.spec.helm != null and .status.summary.ready == 0) | .metadata.namespace + "/" + .metadata.name'
+    record_fail
+}
+
+check_harvester_bundle()
+{
+    echo ">>> Check the Harvester bundle is ready..."
+
+    current_summary=$(mktemp)
+    kubectl get bundles.fleet.cattle.io/mcc-harvester -n fleet-local -o yaml | yq '.status.summary' > $current_summary
+
+    expected_summary=$(mktemp)
+cat > $expected_summary <<EOF
+desiredReady: 1
+modified: 1
+nonReadyResources:
+  - bundleState: Modified
+    modifiedStatus:
+      - apiVersion: v1
+        kind: ConfigMap
+        missing: true
+        name: longhorn-storageclass
+        namespace: longhorn-system
+    name: fleet-local/local
+ready: 0
+EOF
+
+    if ! diff <(yq -P 'sort_keys(..)' $current_summary) <(yq -P 'sort_keys(..)' $expected_summary); then
+        echo "Harvester bundle is not ready!"
+        cat $current_summary
+        record_fail
+        return
+    fi
+
+    echo "The Harvester bundle is ready."
+}
+
+check_nodes()
+{
+    local failed="false"
+    echo ">>> Check all nodes are ready..."
+
+    # Use a file to store the node state becuase we can't set the global variable inside the piped scope
+    # The file is removed in the piped scope if there are not-ready nodes.
+    node_ready_state=$(mktemp)
+
+    # nodes should not be cordoned
+    nodes=$(kubectl get nodes -o yaml)
+    unschedulable=$(echo "$nodes" | yq '.items | any_c(.spec.unschedulable == true)')
+
+    if [ "$unschedulable" = "true" ]; then
+        echo "There are unschedulable nodes:"
+        echo "$nodes" | yq '.items[] | select(.spec.unschedulable == true)  | .metadata.name'
+        rm -f $node_ready_state
+    fi
+
+    # nodes should be ready
+    echo "$nodes" | yq .items[].metadata.name |
+        while read -r node_name; do
+            node_ready=$(kubectl get nodes $node_name -o yaml | yq '.status.conditions | any_c(.type == "Ready" and .status == "True")')
+
+            if [ "$node_ready" = "false" ]; then
+                echo "Node $node_name is not ready!"
+                rm -f $node_ready_state
+            fi
+        done
+
+    if [ -e $node_ready_state ]; then
+        echo "All nodes are ready."
+        rm $node_ready_state
+    else
+        echo "There are non-ready nodes."
+        record_fail
+    fi
+}
+
+check_cluster()
+{
+    echo ">>> Check the CAPI cluster is provisioned..."
+    cluster_phase=$(kubectl get clusters.cluster.x-k8s.io/local -n fleet-local -o yaml | yq '.status.phase')
+
+    # cluster should be in provisioned
+    if [ "$cluster_phase" != "Provisioned" ]; then
+        echo "Cluster is not provisioned ($cluster_phase)"
+        record_fail
+        return
+    fi
+
+    echo "The CAPI cluster is provisioned."
+}
+
+check_machines()
```

Comment:
Might be good to add node count checks? to make sure that there are no orphan machines.

Reply:
Added! https://github.com/harvester/upgrade-helpers/pull/3/commits/545724525ce7f91c1a3aaf62440ea5f05a94811d